modules/transformer-batch/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/batch/PartitionDataFilter.scala

```diff
@@ -0,0 +1,98 @@
+package com.snowplowanalytics.snowplow.rdbloader.transformer.batch
+
+import com.snowplowanalytics.snowplow.rdbloader.common.config.TransformerConfig.Compression
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.Transformed
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FSDataOutputStream, Path}
+import org.apache.hadoop.io.compress.GzipCodec
+
+import java.io.{BufferedWriter, OutputStream, OutputStreamWriter}
+import java.nio.charset.StandardCharsets
+
+object PartitionDataFilter {
+
+  type PartitionData = Iterator[Transformed]
+  type GoodData = Iterator[Transformed]
+  type BadData = Iterator[Transformed]
+
+  def extractGoodAndPersistBad(
+    partitionData: PartitionData,
+    partitionIndex: Int,
+    folderName: String,
+    hadoopConfiguration: Configuration,
+    outputConfig: Config.Output
+  ): GoodData = {
+    val (good, bad) = splitPartitionToGoodAndBad(partitionData)
+
+    if (bad.hasNext) {
+      handleBadrowsFromPartition(folderName, hadoopConfiguration, outputConfig, partitionIndex, bad)
+    }
+
+    good
+  }
+
+  private def splitPartitionToGoodAndBad(data: PartitionData): (GoodData, BadData) =
+    data.partition {
+      case Transformed.WideRow(isGood, _) => isGood
+      case _ => true
+    }
+
+  private def handleBadrowsFromPartition(
+    folderName: String,
+    hadoopConfiguration: Configuration,
+    outputConfig: Config.Output,
+    partitionIndex: Int,
+    badRows: BadData
+  ): Unit = {
+    val outputFile = openFile(folderName, hadoopConfiguration, outputConfig, partitionIndex)
+    val outputStream = createOutputStream(outputFile, hadoopConfiguration, outputConfig)
+    val writer = new BufferedWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8))
+
+    flushBadrowsToFile(badRows, writer)
+    writer.close()
```

Comment:
I've almost forgotten how to write non-effecty (no `F[_]`) scala.  But should this `writer.close()` be in a `finally` block?